pkg/cache/v3/cache.go

```diff
@@ -37,6 +36,27 @@ type Request = discovery.DiscoveryRequest
 // DeltaRequest is an alias for the delta discovery request type.
 type DeltaRequest = discovery.DeltaDiscoveryRequest
 
+// ClientState provides additional data on the client knowledge for the type matching the request
+// This allows proper implementation of stateful aspects of the protocol (e.g. returning only some updated resources)
+// Though the methods may return mutable parts of the state for performance reasons,
+// the cache is expected to consider this state as immutable and thread safe between a watch creation and its cancellation
+type ClientState interface {
+	// GetKnownResources returns the list of resources the clients has ACKed and their associated version.
+	// The versions are:
+	//  - delta protocol: version of the specific resource set in the response
+	//  - sotw protocol: version of the global response when the resource was last ACKed
+	GetKnownResources() map[string]string
+
+	// GetSubscribedResources returns the list of resources currently subscribed to by the client for the type.
+	// For delta it keeps track of subscription updates across requests
+	// For sotw it is a normalized view of the last request resources
+	GetSubscribedResources() map[string]struct{}
```

Comment:
Ah I think I remember why this is needed as a map. Wondering if we that's actually necessary in the cache... I'll take a look

Reply:
The split between `IsWildcard` and `GetSubscribedResources` should now allow to remove the old assumption of "no resources" == "wildcard" which is no longer true. Also a request with `IsWildcard` set to true and explicit resources in `GetSubscribedResources` carries the meaning of "client requested wildcard and explicit resources", and a cache implementation can act on this